lib/srv/alpnproxy/local_proxy.go

```diff
@@ -360,3 +408,59 @@ func (l *LocalProxy) SetCerts(certs []tls.Certificate) {
 	defer l.certsMu.Unlock()
 	l.cfg.Certs = certs
 }
+
+// shouldTunnelConn determines if certificates should be used when dialing
+// upstream to proxy a new downstream connection.
+// After calling shouldTunnelConn function, the returned
+// net.Conn should be used for further operation.
+func (l *LocalProxy) shouldTunnelConn(ctx context.Context, downstreamConn net.Conn) (bool, net.Conn, error) {
+	if l.cfg.Tunnel {
+		return true, downstreamConn, nil
+	}
+	return l.needTunnelConn(ctx, downstreamConn)
+}
+
+func (l *LocalProxy) needTunnelConn(ctx context.Context, downstreamConn net.Conn) (bool, net.Conn, error) {
+	if l.isPostgresProxy() {
+		// `psql` cli doesn't send cancel requests with SSL, unfortunately.
+		// This is a problem when the local proxy has no certs configured,
+		// because normally the client is responsible for connecting with
+		// TLS certificates.
+		// So when the local proxy has no certs configured, we inspect
+		// the connection to see if it is a postgres cancel request and
+		// load certs for the connection.
+		startupMessage, conn, err := peekPostgresStartupMessage(ctx, downstreamConn)
```

Comment:
It's possible and I considered doing something like that, but it would require a lot more refactoring and complexity to do it. FWIW, all of the local proxy changes are due to the requirement that `tsh proxy db` (without `--tunnel` or `--cert-file/--key-file`) loads and uses certs for cancel requests, aka "tunnels on demand". If we instead just let cancel requests fail when client certs are not presented, then all of this extra complexity goes away. I consider it a bug that `psql` doesn't send cancel requests with SSL when I've configured `psql` to connect with client certs.

Reply:
Ya, refactoring can get complicated. The easiest i can think of is to let `middleware.OnNewConnection` return a per connection setting for overwriting downstreamConn and TLS certs. Still a lot of work to change existing middleware but the benefit would be to encapsulate cancel request stuff in a separate middleware that is only used when connecting to Postgres in non-tunnel mode. I'll leave it to you to decide if you are in the mood to refactor in some way or keep it as it is.

regardless, it would be great if we can move pgproto3 stuff to a separate util lib.